File: dataset/CoT/OPENAI_gpt-5-2025-08-07/Task130.java
import java.util.*;

public class Task130 {

    // Response structure
    public static class Response {
        public final int status;
        public final Map<String, String> headers;
        public final String body;

        public Response(int status, Map<String, String> headers, String body) {
            this.status = status;
            this.headers = Collections.unmodifiableMap(new LinkedHashMap<>(headers));
            this.body = body;
        }
    }

    // Security: strict allowlists
    private static final Set<String> ALLOWED_ORIGINS = new HashSet<>(Arrays.asList(
            "https://trustedorigin.com",
            "http://localhost:3000"
    ));
    private static final Set<String> ALLOWED_METHODS = new HashSet<>(Arrays.asList("GET", "POST"));
    private static final Set<String> ALLOWED_HEADERS = new HashSet<>(Arrays.asList(
            "content-type", "authorization", "x-requested-with"
    ));

    // Utility: safe lower-case trim
    private static String normalizeToken(String s) {
        if (s == null) return "";
        return s.trim().toLowerCase(Locale.ROOT);
    }

    private static boolean isOriginAllowed(String origin) {
        if (origin == null) return false;
        // Exact match only to avoid header reflection
        return ALLOWED_ORIGINS.contains(origin.trim());
    }

    private static Set<String> parseHeaderList(String headerList) {
        Set<String> out = new LinkedHashSet<>();
        if (headerList == null) return out;
        String[] parts = headerList.split(",");
        for (String p : parts) {
            String t = normalizeToken(p);
            if (!t.isEmpty()) out.add(t);
        }
        return out;
    }

    private static String joinCSV(Collection<String> items) {
        StringBuilder sb = new StringBuilder();
        boolean first = true;
        for (String it : items) {
            if (!first) sb.append(", ");
            sb.append(it);
            first = false;
        }
        return sb.toString();
    }

    // Core: handle API with CORS
    public static Response handleRequest(
            String method,
            String path,
            String origin,
            String accessControlRequestMethod,
            String accessControlRequestHeaders
    ) {
        String m = method == null ? "" : method.trim().toUpperCase(Locale.ROOT);
        String p = path == null ? "" : path.trim();

        Map<String, String> headers = new LinkedHashMap<>();
        headers.put("X-Content-Type-Options", "nosniff");
        headers.put("X-Frame-Options", "DENY");
        headers.put("Referrer-Policy", "no-referrer");
        headers.put("Cache-Control", "no-store");

        boolean originAllowed = isOriginAllowed(origin);

        if ("OPTIONS".equals(m)) {
            // Preflight
            if (!originAllowed) {
                return new Response(403, headers, "");
            }
            String reqMethod = accessControlRequestMethod == null ? "" :
                    accessControlRequestMethod.trim().toUpperCase(Locale.ROOT);
            if (!ALLOWED_METHODS.contains(reqMethod)) {
                return new Response(405, headers, "");
            }
            Set<String> requested = parseHeaderList(accessControlRequestHeaders);
            List<String> allowedBack = new ArrayList<>();
            if (!requested.isEmpty()) {
                for (String h : requested) {
                    if (ALLOWED_HEADERS.contains(h)) {
                        allowedBack.add(h);
                    }
                }
            } else {
                allowedBack.addAll(ALLOWED_HEADERS);
            }
            // CORS headers
            headers.put("Access-Control-Allow-Origin", origin.trim());
            headers.put("Vary", "Origin");
            headers.put("Access-Control-Allow-Methods", joinCSV(ALLOWED_METHODS));
            if (!allowedBack.isEmpty()) {
                headers.put("Access-Control-Allow-Headers", joinCSV(allowedBack));
            }
            headers.put("Access-Control-Max-Age", "600");
            // Do not set Allow-Credentials unless needed; omitting prevents inadvertent exposure
            return new Response(204, headers, "");
        }

        // Actual request
        if (!"/api/data".equals(p)) {
            return new Response(404, headers, "{\"error\":\"not found\"}");
        }
        if (!ALLOWED_METHODS.contains(m)) {
            return new Response(405, headers, "{\"error\":\"method not allowed\"}");
        }

        if (originAllowed) {
            headers.put("Access-Control-Allow-Origin", origin.trim());
            headers.put("Vary", "Origin");
            headers.put("Access-Control-Expose-Headers", "Content-Type, Content-Length");
            // Not setting Allow-Credentials by default for safety
        }
        headers.put("Content-Type", "application/json; charset=utf-8");

        String body = "{\"message\":\"Hello from CORS protected API\",\"path\":\"" + p + "\"}";
        return new Response(200, headers, body);
    }

    // Test helper: pretty print a response (for demonstration)
    private static void printResponse(String title, Response r) {
        System.out.println("== " + title + " ==");
        System.out.println("Status: " + r.status);
        for (Map.Entry<String, String> e : r.headers.entrySet()) {
            System.out.println(e.getKey() + ": " + e.getValue());
        }
        System.out.println("Body: " + r.body);
        System.out.println();
    }

    public static void main(String[] args) {
        // 5 test cases
        Response t1 = handleRequest(
                "OPTIONS",
                "/api/data",
                "https://trustedorigin.com",
                "GET",
                "Content-Type, X-Requested-With"
        );
        Response t2 = handleRequest(
                "OPTIONS",
                "/api/data",
                "https://evil.com",
                "GET",
                "Content-Type"
        );
        Response t3 = handleRequest(
                "GET",
                "/api/data",
                "http://localhost:3000",
                null,
                null
        );
        Response t4 = handleRequest(
                "OPTIONS",
                "/api/data",
                "http://localhost:3000",
                "DELETE",
                "Content-Type"
        );
        Response t5 = handleRequest(
                "GET",
                "/api/data",
                "https://unknown.com",
                null,
                null
        );

        printResponse("Test1 Allowed Origin Preflight", t1);
        printResponse("Test2 Disallowed Origin Preflight", t2);
        printResponse("Test3 Allowed Origin Actual GET", t3);
        printResponse("Test4 Not Allowed Method Preflight", t4);
        printResponse("Test5 Disallowed Origin Actual GET", t5);
    }
}

File: dataset/CoT/OPENAI_gpt-5-2025-08-07/Task130.py
class Response:
    def __init__(self, status: int, headers: dict, body: str):
        self.status = status
        self.headers = dict(headers)
        self.body = body

# Security: allowlists
ALLOWED_ORIGINS = {
    "https://trustedorigin.com",
    "http://localhost:3000",
}
ALLOWED_METHODS = {"GET", "POST"}
ALLOWED_HEADERS = {"content-type", "authorization", "x-requested-with"}

def _normalize_token(s: str) -> str:
    return (s or "").strip().lower()

def _is_origin_allowed(origin: str) -> bool:
    if not origin:
        return False
    return origin.strip() in ALLOWED_ORIGINS

def _parse_header_list(header_list: str) -> list:
    if not header_list:
        return []
    return [t.strip().lower() for t in header_list.split(",") if t.strip()]

def _join_csv(items) -> str:
    return ", ".join(items)

def handle_request(method: str, path: str, origin: str,
                   access_control_request_method: str = None,
                   access_control_request_headers: str = None) -> Response:
    m = (method or "").strip().upper()
    p = (path or "").strip()

    headers = {
        "X-Content-Type-Options": "nosniff",
        "X-Frame-Options": "DENY",
        "Referrer-Policy": "no-referrer",
        "Cache-Control": "no-store",
    }

    origin_allowed = _is_origin_allowed(origin)

    if m == "OPTIONS":
        if not origin_allowed:
            return Response(403, headers, "")
        req_method = (access_control_request_method or "").strip().upper()
        if req_method not in ALLOWED_METHODS:
            return Response(405, headers, "")
        requested = _parse_header_list(access_control_request_headers)
        allowed_back = []
        if requested:
            for h in requested:
                if h in ALLOWED_HEADERS:
                    allowed_back.append(h)
        else:
            allowed_back = list(ALLOWED_HEADERS)

        headers["Access-Control-Allow-Origin"] = origin.strip()
        headers["Vary"] = "Origin"
        headers["Access-Control-Allow-Methods"] = _join_csv(sorted(ALLOWED_METHODS))
        if allowed_back:
            headers["Access-Control-Allow-Headers"] = _join_csv(sorted(allowed_back))
        headers["Access-Control-Max-Age"] = "600"
        return Response(204, headers, "")

    if p != "/api/data":
        return Response(404, headers, '{"error":"not found"}')
    if m not in ALLOWED_METHODS:
        return Response(405, headers, '{"error":"method not allowed"}')

    if origin_allowed:
        headers["Access-Control-Allow-Origin"] = origin.strip()
        headers["Vary"] = "Origin"
        headers["Access-Control-Expose-Headers"] = "Content-Type, Content-Length"
    headers["Content-Type"] = "application/json; charset=utf-8"

    body = '{"message":"Hello from CORS protected API","path":"' + p + '"}'
    return Response(200, headers, body)

def _print_response(title: str, r: Response):
    print(f"== {title} ==")
    print(f"Status: {r.status}")
    for k, v in r.headers.items():
        print(f"{k}: {v}")
    print(f"Body: {r.body}\n")

def main():
    # 5 test cases
    t1 = handle_request(
        "OPTIONS", "/api/data", "https://trustedorigin.com", "GET", "Content-Type, X-Requested-With"
    )
    t2 = handle_request(
        "OPTIONS", "/api/data", "https://evil.com", "GET", "Content-Type"
    )
    t3 = handle_request(
        "GET", "/api/data", "http://localhost:3000"
    )
    t4 = handle_request(
        "OPTIONS", "/api/data", "http://localhost:3000", "DELETE", "Content-Type"
    )
    t5 = handle_request(
        "GET", "/api/data", "https://unknown.com"
    )

    _print_response("Test1 Allowed Origin Preflight", t1)
    _print_response("Test2 Disallowed Origin Preflight", t2)
    _print_response("Test3 Allowed Origin Actual GET", t3)
    _print_response("Test4 Not Allowed Method Preflight", t4)
    _print_response("Test5 Disallowed Origin Actual GET", t5)

if __name__ == "__main__":
    main()